src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -550,12 +550,13 @@ inline void ExceptionMark::check_no_pending_exception() {
   }
 }
 
+extern bool is_vm_created();
 
 ExceptionMark::~ExceptionMark() {
   if (_thread->has_pending_exception()) {
     Handle exception(_thread, _thread->pending_exception());
     _thread->clear_pending_exception(); // Needed to avoid infinite recursion
-    if (is_init_completed()) {
+    if (is_vm_created()) {
```

Comment:
Can you explain this change please.

Reply:
BTW relying on the `ExceptionMark` to catch and report an OOME that is (to a degree) expected, seems like a misuse of EM to me. I don't know which EM is involved but perhaps the right fix there is to not have the EM and instead check for the exception and "return" and let the launcher report the OOME directly?

Reply:
> I think the thought process here was that we should "never" get exceptions during these parts of VM init because the minimum heap size should ensure the VM can safely init. Hence getting an actual exception was indeed a "fatal error" because something was misconfigured - hence the EM. With AoTOC that may no longer be the case.

Yeah I suspected that was the case. And that previously made sense - throwing would most likely be a bug in the JVM before. But with AOT objects, we no longer have such a minimal heap size, and it is unlikely that the cause of the OOME is a bug in the JVM, and much more likely that the heap was too small to accomodate the AOT cache objects.

Reply:
The AOT cache knows what objects (and their individual sizes) are archived. However, with streaming, the exact heap size needed to materialize all the archived objects depends on the GC type and GC tuning parameters. For example, a region-based GC could segregate objects by their sizes, so there will be an overhead (unused space in the segregated regions) that's hard to calculate unless you do a pre-scan of all the archived objects.